src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,201 @@
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography;
+
+using CURLoption = Interop.libcurl.CURLoption;
+using CURLcode  = Interop.libcurl.CURLcode;
+using CURLINFO = Interop.libcurl.CURLINFO;
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+
+        private sealed class ClientCertProvider : IDisposable
+        {
+            private static readonly Interop.libcurl.ssl_ctx_callback s_sslCallback = CurlSslCallback;
+            private string _certificateFile;
+            private X509Certificate2 _certificate;
+
+            private ClientCertProvider(X509Certificate2 certificate)
+            {
+                _certificate = certificate;
+            }
+
+            private void SetCallbacks(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.Callback);
+                Debug.Assert(_certificate != null);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_FUNCTION, s_sslCallback);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_DATA, _certificate.Handle);
+                VerboseTrace("Set Callback");
+            }
+
+            private void SetClientCertCommonName(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.CommonName);
+                string commonName = GetCommonName(_certificate);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERT, commonName);
+                VerboseTrace("Set CommonName");
+            }            
+
+            private void ExportToFile(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.FileExport);
+
+                string temporaryDirectory = PersistedFiles.GetTempFeatureDirectory("cryptography");
+                Directory.CreateDirectory(temporaryDirectory);
+                _certificateFile = Path.Combine(temporaryDirectory, Guid.NewGuid().ToString("N"));
+
+                string password = GeneratePassword();
+                byte[] pfxData = _certificate.Export(X509ContentType.Pkcs12, password);
+                File.WriteAllBytes(_certificateFile, pfxData);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERT, _certificateFile);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSLCERTTYPE, CertificateFileType);
+                easy.SetCurlOption(CURLoption.CURLOPT_KEYPASSWD, password);
+                VerboseTrace(_certificateFile + " created");
+            }
+
+            internal void SetClientCertificateOptions(EasyRequest easy)
+            {
+                Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.Callback ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.CommonName ||
+                             s_clientCertificatePolicy == ClientCertificatePolicy.FileExport);
```

Comment:
Rather than this assert, you could just do Debug.Assert(s_clientCertificatePolicy == ClientCertificatePolicy.FileExport) in the default case of the switch.